tools/cloud-build/babysit_tests.py

```diff
@@ -0,0 +1,233 @@
+
+    def _approve(self, build: Build) -> None:
+        req = ApproveBuildRequest(
+            name=f"projects/{build.project_id}/builds/{build.id}",
+            approval_result=ApprovalResult(
+                decision=ApprovalResult.Decision.APPROVED
+            )
+        )
+        self.cb.approve_build(request=req)
+
+    def _retry(self, build: Build) -> None:
+        req = RetryBuildRequest(project_id=build.project_id, id=build.id)
+        self.cb.retry_build(request=req)
+
+
+    def _act(self, builds: Sequence[Build]) -> bool:
+        latest = latest_by_trigger(builds).values()
+        active = [bc.build for bc in latest if not self._in_terminal_state(bc)]
+        if not active:
+            return False  # all builds are in terminal state, done
+
+        not_running = [b for b in active if b.status not in (
+            Status.QUEUED, Status.WORKING)]
+        num_running = len(active) - len(not_running)
+
+        if num_running == len(active):
+            return True  # waiting for results
+        if num_running >= self.concurrency:
+            return True  # waiting for "openning"
+
+        pend = next(
+            (b for b in not_running if b.status == Status.PENDING), None)
+        if pend is not None:  # approve one of pending builds
+            self._approve(pend)
+            return True
+
+        assert not_running # sanity check
+        failed = not_running[0]
+        assert failed.status in [Status.FAILURE, Status.INTERNAL_ERROR, Status.TIMEOUT]  # sanity check
+        self._retry(failed) # retry failed build
+        return True
+        
+
+    def _sleep(self) -> None:
+        time.sleep(5)
+        print(".", end="", flush=True)
+
+    def _render_summary(self, builds: Sequence[Build]) -> None:
+        for _, bc in sorted(latest_by_trigger(builds).items()):
+            print(render_build(bc.build, bc.count))
+
+    def _track_status_updates(self, builds: Sequence[Build]) -> None:
+        if not self._status: # state is empty, just populate
+            for b in builds:
+                self._status[b.id] = b.status
+            return
+        
+        for b in builds:
+            if b.status != self._status.get(b.id):
+                print_line(f"{render_status(self._status.get(b.id))} > {render_build(b)}")
+                self._status[b.id] = b.status
+
+
+    def do(self):
+        builds = self._get_builds()
+        if not builds:
+            print(f"Found no builds referencing SHA {self.sha}")
+            return
+
+        print(f"Found {len(builds)} builds:")
+        self._render_summary(builds)
+        
+        acted = False
+        while self._act(builds):
+            acted = True
+            self._sleep()
+            builds = self._get_builds()
+            self._track_status_updates(builds)
+
+        print("Done, no actions left to take.")
+        if acted:
+            self._render_summary(builds)
+
+
+def render_status(status):
```

Comment:
What is the rational to keep these three functions (render_status, render_build, print_line) outside of class even though they are not called by anything else?

Reply:
See a new version of PR. All UI was refactored into a dedicated class.